js/ui/watermark.js

```diff
@@ -0,0 +1,237 @@
+        this._settings.connect('changed::watermark-border',
+                               Lang.bind(this, this._updateBorder));
+        this._settings.connect('changed::watermark-always-visible',
+                               Lang.bind(this, this._updateVisibility));
+
+        this._textureCache = St.TextureCache.get_default();
+        this._textureCache.connect('texture-file-changed', Lang.bind(this,
+            function(cache, file) {
+                if (!this._watermarkFile || !this._watermarkFile.equal(file))
+                    return;
+
+                this._updateWatermarkTexture();
+            }));
+
+        this.actor = new St.Widget({ layout_manager: new Clutter.BinLayout(),
+                                     opacity: 0 });
+        bgManager._container.add_actor(this.actor);
+
+        this.actor.connect('destroy', Lang.bind(this, this._onDestroy));
+
+        let monitorIndex = bgManager._monitorIndex;
+        let constraint = new Monitor.MonitorConstraint({ index: monitorIndex,
+                                                         work_area: true});
+        this.actor.add_constraint(constraint);
+
+        this._bin = new St.Widget({ x_expand: true, y_expand: true });
+        this.actor.add_actor(this._bin);
+
+        this._settings.bind('watermark-opacity', this._bin, 'opacity',
+                            Gio.SettingsBindFlags.DEFAULT);
+        this._settings.connect('changed::watermark-opacity',
+                               Lang.bind(this, this._updateOpacity));
+
+        this._updateWatermark();
+        this._updatePosition();
+        this._updateOpacity();
+        this._updateBorder();
+
+        this._bgDestroyedId =
+            bgManager.backgroundActor.connect('destroy',
+                                              Lang.bind(this, this._backgroundDestroyed));
+
+        this._bgChangedId =
+            bgManager.connect('changed', Lang.bind(this, this._updateVisibility));
+
+        this._updateVisibility();
+    },
+
+    _updateOpacity: function() {
+        log('opacity changed to ' + this._settings.get_uint('watermark-opacity'));
+    },
+
+    _updateWatermark: function() {
+        let filename = this._settings.get_string('watermark-file');
+        let file = Gio.File.new_for_commandline_arg(filename);
+        if (this._watermarkFile && this._watermarkFile.equal(file))
+            return;
+
+        this._watermarkFile = file;
+
+        this._updateWatermarkTexture();
+    },
+
+    _updateWatermarkTexture: function() {
+        if (this._icon)
+            this._icon.destroy();
+
+        let scaleFactor = St.ThemeContext.get_for_stage(global.stage).scale_factor;
+
+        log('updating watermark texture');
+
+        this._icon = this._textureCache.load_file_async(this._watermarkFile, -1, -1, scaleFactor);
+        this._icon.connect('allocation-changed', Lang.bind(this, this._updateScale));
+        this._bin.add_actor(this._icon);
+        this._bin.queue_redraw();
```

Comment:
Not sure you need to manually queue a draw here; I'd imagine that adding the new actor will already cause that to happen